### PR TITLE
feat: depth estimation dataset guide.

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR adds a guide for prepping datasets for depth estimation. 

PR to add documentation images is up here: https://huggingface.co/datasets/huggingface/documentation-images/discussions/22

### CODE REVIEW

File: docs/source/depth_estimation.mdx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+pip install -U albumentations 
+```
+
+[Albumentations](https://albumentations.ai/) is a Python library for performing data augmentation
+for computer vision. It supports various computer vision tasks such as image classification, object
+detection, segmentation, and keypoint estimation.
+
+This guide uses the [NYU Depth V2](https://huggingface.co/datasets/sayakpaul/nyu_depth_v2) dataset which is 
+comprised of video sequences from a variety of indoor scenes as recorded by both the RGB and Depth cameras
+from the Microsoft [Kinect](http://www.xbox.com/kinect). The dataset consists of scenes from 3 cities and provides images along with
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:20]: I don't think it's necessary or relevant to mention here that it was recorded from a Kinect. A link to the dataset card is fine :)

```suggestion
comprised of video sequences from various indoor scenes, recorded by RGB and depth cameras. The dataset consists of scenes from 3 cities and provides images along with
```

File: docs/source/depth_estimation.mdx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+>>> from datasets import load_dataset
+
+>>> train_dataset = load_dataset("sayakpaul/nyu_depth_v2", split="train")
+>>> index = 17
+>>> dataset[index]
+{'image': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=640x480>,
+ 'depth_map': <PIL.PngImagePlugin.PngImageFile image mode=L size=640x480>}
+```
+
+The dataset has three fields:
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:35]:
```suggestion
The dataset has two fields:
```

File: docs/source/depth_estimation.mdx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+```py
+>>> train_dataset[index]["image"]
+```
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_sample.png">
+</div>
+
+Similarly, you can check out the respective depth map:
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:50]: Probably better if we tell the user to look at the depth map instead of suggesting it

```suggestion
Now take a look at its corresponding depth map:
```

File: docs/source/depth_estimation.mdx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+```py
+>>> train_dataset[index]["depth_map"]
+```
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_target.png">
+</div>
+
+It's all black! You need to apply some preprocessing to the depth map to properly visualize it.
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:60]: I would delete the **note** and combine the code snippets. 

```suggestion
It's all black! You'll need to add some color to the depth map to visualize it properly (the utility below is taken from the [FastDepth repository](https://github.com/dwofk/fast-depth/blob/master/utils.py)).
```

File: docs/source/depth_estimation.mdx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+Since these transformations (such as flipping, cropping) also impact the depth maps, you need to define a mapping
+while applying these transformations:
+
+```py 
+>>> additional_targets = {"depth": "mask"}
+>>> aug = A.Compose(transforms=transforms, additional_targets=additional_targets)
+```
+
+This will ensure that you can pass an image and its corresponding depth map to `aug`.
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:125]:
```suggestion
This ensures you can pass an image and its corresponding depth map to `aug`.
```

File: docs/source/depth_estimation.mdx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+...     return examples
+```
+
+Use the [`~Dataset.set_transform`] function to apply the transformation on-the-fly to batches of the dataset to consume less disk space:
+
+```py
+>>> train_dataset.set_transform(transforms)
+```
+
+You can verify the transformation worked by indexing into the `pixel_values` and `labels` of an example:
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:149]: I would combine the two code snippets below.

```suggestion
You can verify the transformation worked by indexing into the `pixel_values` and `labels` of an example image:
```

File: docs/source/depth_estimation.mdx [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+>>> plt.imshow(image)
+>>> plt.axis("off")
+>>> plt.show()
+```
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_sample_aug.png">
+</div>
+
+```py
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:166]: Maybe add some text here that says something like:

`Visualize the same transformation on the image's corresponding depth map:`

File: docs/source/depth_estimation.mdx (hunk 6 above)
Comment (sayakpaul) [reply docs/source/depth_estimation.mdx]: The succession made it visually simpler to read through. Hence decided to keep it that way.

File: docs/source/depth_estimation.mdx (hunk 4 above)
Comment (sayakpaul) [reply docs/source/depth_estimation.mdx]: See if it looks better now.

File: docs/source/depth_estimation.mdx (hunk 1 above)
Comment (sayakpaul) [reply docs/source/depth_estimation.mdx]: Agreed. Thanks.

**Comment by stevhliu** [comment]
Thanks for the changes, looks good to me!

**Comment by sayakpaul** [comment]
@stevhliu I have pushed some quality improvements both in terms of code and content. Would you be able to re-review?

File: docs/source/depth_estimation.mdx [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_collage.png">
+</div>
+
+Now apply some augmentations with `albumentations`. The augmentation transformations include:
+
+* Random cropping 
+* Random brightness and contrast 
+* Random gamma correction 
+* Random hue saturation
```
Comment (sayakpaul) [inline docs/source/depth_estimation.mdx:130]: I had to drop horizontal flipping. Here's why. 

The behavior seems to be unstable with `datasets` when applied via `set_transform()`. Consider the following as an example. 

debug_augmentations

You'll notice that subplots tagged as "wrong" have flipped images but their corresponding depth maps are not flipped. This is not a consistent case, though. Subplots tagged as "right", for example. You'll notice that the images and their depth maps have both been flipped there (compare them to the originals and you should be able to spot the difference). 

I have verified if there's a problem with `albumentations` but that doesn't seem to be the case. Refer to this notebook: https://gist.github.com/sayakpaul/63ae5f7971f634894ab37e76e6b3879c. You'll notice that I took individual samples from the dataset and passed them through the `aug` chain and the results were fine. 

Have you experienced this before? Is anything obviously wrong you can spot in the code? 

Let me know if anything's unclear. 

@nateraw @lhoestq

File: docs/source/depth_estimation.mdx [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+...    plt.show()
+
+>>> show_depthmap(train_dataset[index]["depth_map"])
+```
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_target_viz.png">
+</div>
+
+You can also visualize a collage of different images and the corresponding depth maps.
```
Comment (stevhliu) [inline docs/source/depth_estimation.mdx:96]: I think the wording might be a little clearer this way. Typically `collage` has a very specific reference to art. 

```suggestion
You can also visualize several different images and their corresponding depth maps.
```

**Review by stevhliu — APPROVED** [review]
Very cool, love the new `show_depthmap` function! :)

File: docs/source/depth_estimation.mdx (hunk 8 above)
Comment (nateraw) [reply docs/source/depth_estimation.mdx:130]: I'm unable to reproduce this issue 🙁 . Take a look at this [simplified notebook](https://gist.github.com/nateraw/434fd94de897d76f449a805fcc3ec623).

Comment (sayakpaul) [reply docs/source/depth_estimation.mdx:130]: Yes @nateraw. But it doesn't seem to be consistent, as I stated here:

> You'll notice that subplots tagged as "wrong" have flipped images but their corresponding depth maps are not flipped. This is not a consistent case, though. Subplots tagged as "right", for example.

Sometimes, the augmented depth maps are correct, i.e., the flip transformation has been correctly applied, and sometimes, that doesn't seem to be the case. You'll notice in the collage [here](https://github.com/huggingface/datasets/pull/5379#discussion_r1056065300).

Comment (nateraw) [reply docs/source/depth_estimation.mdx:130]: Yea I noticed that, but I ran it over and over and never got that behavior. Do you have a minimal reproduction of the behavior?

Comment (sayakpaul) [reply docs/source/depth_estimation.mdx:130]: I don't. Strangely, it only comes up when I try to put it inside a collection of images.

I will try to reproduce the behavior minimally and let you know here.

Comment (sayakpaul) [reply docs/source/depth_estimation.mdx:130]: My god! Thanks for saving me!

**Comment by sayakpaul** [comment]
@lhoestq @nateraw made some changes as per the comments. PTAL and approve as necessary.

**Review by lhoestq — APPROVED** [review]
Awesome ! LGTM :)

**Comment by sayakpaul** [comment]
Merging this PR with approvals from @stevhliu @lhoestq.
